build/patches/Enable-crash-reporting-ui.patch

```diff
@@ -43,30 +162,131 @@ diff --git a/chrome/browser/ui/webui/crashes_ui.cc b/chrome/browser/ui/webui/cra
  
    bool system_crash_reporter = false;
  #if BUILDFLAG(IS_CHROMEOS_ASH) || BUILDFLAG(IS_CHROMEOS_LACROS)
-@@ -244,11 +243,6 @@ void CrashesDOMHandler::HandleRequestSingleCrashUpload(
+@@ -244,12 +261,78 @@ void CrashesDOMHandler::HandleRequestSingleCrashUpload(
    bool success = args->GetString(0, &local_id);
    DCHECK(success);
  
 -  // Only allow manual uploads if crash uploads aren’t disabled by policy.
 -  if (!ChromeMetricsServiceAccessor::IsMetricsAndCrashReportingEnabled() &&
 -      IsMetricsReportingPolicyManaged()) {
 -    return;
--  }
-   upload_list_->RequestSingleUploadAsync(local_id);
++  base::ThreadPool::PostTaskAndReplyWithResult(
++      FROM_HERE, kLoadingTaskTraits,
++      base::BindOnce(&CrashesDOMHandler::RequestSingleUpload, base::Unretained(this), local_id),
++      base::BindOnce(&CrashesDOMHandler::RequestSingleUploadCallback, base::Unretained(this), local_id));
++}
++
++std::string CrashesDOMHandler::RequestSingleUpload(const std::string& local_id) const {
++  // get crash file path
++  std::string info_file_path = upload_list_->GetFilePathByLocalId(local_id);
++  if (info_file_path.empty()) {
++    LOG(ERROR) << "Crash report: file path is not set for " << local_id;
++    return std::string();
++  }
++  base::FilePath crash_file_path(info_file_path);
++
++  // get android crash report dir
++  base::FilePath cache_dir;
++  base::android::GetCacheDirectory(&cache_dir);
++  base::FilePath upload_log_path = cache_dir.Append("Crash Reports");
++
++  // crash reports can have multiple extensions (e.g. foo.dmp, foo.dmp.try1,
++  // foo.skipped.try0), remove it
++  base::FilePath zip_file = crash_file_path;
++  while (zip_file != zip_file.RemoveExtension())
++    zip_file = zip_file.RemoveExtension();
++
++  // make zip file name, like "ec708a7b-cb17-44e7-8dae-e32f6c45cb8c.zip"
++  zip_file = upload_log_path.Append(zip_file.BaseName())
++                            .AddExtensionASCII(".zip");
++  // if exists return it
++  if (base::PathExists(zip_file))
++    return zip_file.value();
++
++  // original code remove the file immediately after upload.
++  // we changed this behavior but it is still possible that the file no longer exists
++  // because in uploads.log it could be indicated but the file was deleted by self-cleaning
++  if (!base::PathExists(crash_file_path)) {
++    LOG(ERROR) << "Crash report: file " << crash_file_path
++               << " no more avaiable";
++    return std::string();
++  }
++
++  // temporary folder for zip, auto-removed when out of scope
++  base::ScopedTempDir temp_dir;
++  if (!temp_dir.CreateUniqueTempDir()) {
++    LOG(ERROR) << "Crash report: cannot create temp directory";
++    return std::string();
++  }
++
++  // temporary file for zip content
++  base::FilePath dest_file = temp_dir.GetPath().Append(crash_file_path.BaseName());
++
++  // make zip file
++  if (base::CopyFile(crash_file_path, dest_file)) {
++    if (zip::Zip(temp_dir.GetPath(), zip_file, false)) {
```

Comment:
I create a temporary directory and put all the files I want in the zip.
that one called `zip::Zip()` is the only one I found in all the native side chromium code.
it needs a source folder and a destination file, there are no other methods

Reply:
There is `zip::ZipFiles`

Reply:
no need to change it, if you look in the chromium code they all do exactly as I wrote

Reply:
I am not reviewing Chromium code, I am reviewing the code in this patch: is the file copy necessary?

Reply:
>  is the file copy necessary?

yes, there is no other way

> There is zip::ZipFiles

it's under a if-define and never used

```
#if defined(OS_POSIX)
bool ZipFiles(const base::FilePath& src_dir,
              const std::vector<base::FilePath>& src_relative_paths,
              int dest_fd) {
```

Reply:
> OS_POSIX is defined for Android.

ah, ok, I didn't know

>> yes, there is no other way

> Why?

maybe I don't understand the matter.
this is the header
```
// Convenience method for callers who don't need to set up the filter callback.
// If |include_hidden_files| is true, files starting with "." are included.
// Otherwise they are omitted.
bool Zip(const base::FilePath& src_dir, const base::FilePath& dest_file,
         bool include_hidden_files);

#if defined(OS_POSIX)
// Zips files listed in |src_relative_paths| to destination specified by file
// descriptor |dest_fd|, without taking ownership of |dest_fd|. The paths listed
// in |src_relative_paths| are relative to the |src_dir| and will be used as the
// file names in the created zip file. All source paths must be under |src_dir|
// in the file system hierarchy.
bool ZipFiles(const base::FilePath& src_dir,
              const std::vector<base::FilePath>& src_relative_paths,
              int dest_fd);
#endif  // defined(OS_POSIX)
```
unlike the other, `ZipFiles` is not used anywhere in the code
![image](https://user-images.githubusercontent.com/29201891/117937230-09c4cc00-b306-11eb-9921-4b77302cd255.png)

if I can I prefer code already used, even just to see how to use it.
what I do not understand?

Reply:
I know it is not used currently in Chromium, but it is available.

You have still not explained why the file copy is necessary.

Reply:
> You have still not explained why the file copy is necessary.

also for the other patch you raised the same question.
why a copy in a temporary directory is bothering you, considering by the way that the entire chromium base code does that way? 

really, it is not the will not to do it, it is that I cannot understand the motivation.

Reply:
@uazo the file copy is not bothering me, your attitude is. You have not explained why the file copy is necessary and are being evasive.

Motivation: if it is not necessary it should not be there.

Reply:
> You have not explained why the file copy is necessary and are being evasive.

evasive? I have already written you my reasons:

1. I have not found any other method to zip
2. It's used that way in all chromium
3. The function you indicated is used only in tests

permettimi in italiano: nello sviluppo di chromium mi baso sul framework che danno a disposizione, la ragione tecnica del fatto che zip() richieda una cartella in input non la conosco, ma quel metodo lo richiede, ed è fatto così. 

> Motivation: if it is not necessary it should not be there.

zipping the crash file was my proposal to remedy any encoding problems once posted on github.
if you think it is not a problem, we can keep the file unzipped.

Reply:
> 1. I have not found any other method to zip

I have told you that there is another method.

> 2\. It's used that way in all chromium

It does not matter.

> 3\. The function you indicated is used only in tests

It does not matter.

You have replied:
> > is the file copy necessary?
> 
> yes, there is no other way

And I showed you another way. The fact that it is done like that in Chromium or the method is currently used only in tests does not mean that there is no other way, that is your conclusion that I do not agree with.

If you do not want to change it, no worries, I will do it later.

Reply:
>  The fact that it is done like that in Chromium or the method is currently used only in tests does not mean that there is **no other way**

that answer I had given before knowing that OS_POSIX is active.
later, checking, I saw that it is only used in tests and we have avoided the use of test methods other times as well.

> If you do not want to change it, no worries, I will do it later.

it's not that, in my opinion it's not worth it, but of course do what you think.

Reply:
> we have avoided the use of test methods other times as well.

This is not a test method. It is a method currently used only in tests, it is different. It has not been developed for using it only in tests (it does not have "test" mentioned in the name or description).